### PR TITLE
Intel switchdev

### PR DESCRIPTION
**Pull request by zeeke**
Ice driver supports creating VFs after the eSwitcMode is
set to switchdev. This is different than the preferred way
for `mlx5` driver.

Add specific functions to configure eSwitchMode and SriovNumVFs
for each driver.

Add a basic test to loop over all available
devices and test if setting `eSwitchMode: switchdev` works.

### CODE REVIEW

**Review by ykulazhenkov — DISMISSED** [review]
LGTM

**Comment by Eoghan1232** [comment]
I will do a test of this

**Comment by Eoghan1232** [comment]
worked like a charm @zeeke  ;) 

let me give the PR a review now then we should be good to merge

**Review by Eoghan1232 — APPROVED** [review]
Changes LGTM!

once this is merged, we should also update the support HW list.
I will do this

File: pkg/host/internal/sriov/sriov.go
Comment (SchSeba) [inline pkg/host/internal/sriov/sriov.go:995]: nit: can we move this one outside the function?

File: pkg/host/internal/sriov/sriov.go
Comment (SchSeba) [inline pkg/host/internal/sriov/sriov.go:997]: nit: can we do something like

```
setEswitchModeAndNumVFsByDriverName map[string]setEswitchModeAndNumVFsFn := map[string]setEswitchModeAndNumVFsFn{...}
```

File: pkg/host/internal/sriov/sriov.go
Comment (SchSeba) [inline pkg/host/internal/sriov/sriov.go:990]: nit: can we please add a debug log level about the PCI address and the driver name?

File: pkg/host/internal/sriov/sriov.go
Comment (SchSeba) [inline pkg/host/internal/sriov/sriov.go:1005]: can you please add a debug message that we didn't find the driver in the supported list so we fallback

File: pkg/host/internal/sriov/sriov.go
Comment (SchSeba) [inline pkg/host/internal/sriov/sriov.go:1011]: please add a function comment to explain the process of how we configure switchDev virtual functions for this case

File: pkg/host/internal/sriov/sriov.go
Comment (SchSeba) [inline pkg/host/internal/sriov/sriov.go:1034]: please add a function comment to explain the process of how we configure switchDev virtual functions for this case

File: test/conformance/tests/test_switchdev.go
Comment (SchSeba) [inline test/conformance/tests/test_switchdev.go:25]: nit: please update the skip message here :)

File: test/conformance/tests/test_switchdev.go
Comment (SchSeba) [inline test/conformance/tests/test_switchdev.go:47]: you can use `findUnusedSriovDevices` to have the list of non main device :)

File: pkg/host/internal/sriov/sriov.go
Comment (zeeke) [reply pkg/host/internal/sriov/sriov.go:995]: ok, moving. though that type is not supposed to be used out of this function

File: pkg/host/internal/sriov/sriov.go
Comment (zeeke) [reply pkg/host/internal/sriov/sriov.go:1005]: good point!

**Comment by adrianchiris** [comment]
Please hold off with merging this one id like to go over this pr by end of week

File: pkg/host/internal/sriov/sriov.go
Comment (Eoghan1232) [inline pkg/host/internal/sriov/sriov.go:1051]: we need to reset vf's to 0 before setting switchdev mode, is this handled?

File: pkg/host/internal/sriov/sriov.go
Comment (adrianchiris) [reply pkg/host/internal/sriov/sriov.go:1024]: nit: between b. and c. we also unbind all VFs on PF.
can you add this to the comment ?

note: rebind happens after part of `configSriovVFDevices`

also what about `configureHWOptionsForSwitchdev` ? is it the same for intel ice ?

**Review by adrianchiris — COMMENTED** [review]
overall lgtm, left one small comment

File: pkg/host/internal/sriov/sriov.go
Comment (Eoghan1232) [reply pkg/host/internal/sriov/sriov.go]: I believe so, otherwise driver will throw an error I believe. 
I can double check this tomorrow to confirm

Comment (Eoghan1232) [reply pkg/host/internal/sriov/sriov.go]: I can see in some docs that we must have 0 VF's before switching to switchdev mode.

File: pkg/host/internal/sriov/sriov.go
Comment (Eoghan1232) [reply pkg/host/internal/sriov/sriov.go:1025]: I don't think we need to unbind driver, since we need to reset back to 0 vf's anyways

Comment (zeeke) [reply pkg/host/internal/sriov/sriov.go:1025]: Got it. Then better move the `unbindAllVFsOnPF(...)` call outside of `setEswitchMode(...)`

Comment (adrianchiris) [reply pkg/host/internal/sriov/sriov.go:1025]: > Regarding configureHWOptionsForSwitchdev, current implementation is about setting flow_steering_mode=smfs, which is Mellanox only (IIUC).

this also enables hw-tc-offload via ethtool.

Comment (zeeke) [reply pkg/host/internal/sriov/sriov.go:1025]: > > Regarding configureHWOptionsForSwitchdev, current implementation is about setting flow_steering_mode=smfs, which is Mellanox only (IIUC).
> 
> this also enables hw-tc-offload via ethtool.

right! I miss that. I checked and it should not be a problem for the ice driver:

from a local test: 
```
2024-08-29T08:21:49.27421628Z   LEVEL(-2)       sriov/sriov.go:365      EnableHwTcOffload(): enable offloading  {"device": "eno12409"}
2024-08-29T08:21:49.274526639Z  LEVEL(-2)       sriov/sriov.go:365      EnableHwTcOffload(): already enabled    {"device": "eno12409"}
```

and `flow_steering_mode` setting is guarded: if it's not supported, it doesn't hurt.

**Review by SchSeba — APPROVED** [review]
LGTM

nice work!

**Review by e0ne — APPROVED** [review]

**Comment by zeeke** [comment]
Got approval from multiple sides, merging.

**Pull request merged**
